### DISCUSSION TITLE
Standalone styled-components transpiler

### DISCUSSION THREAD

**Discussion by Xiphe** [styled-components/styled-components#3789] [Ideas]
As we're currently moving into a next generation of tooling and it's not save to assume that most projects use babel anymore,
I wonder If it might be valuable to have a new transpiler doing everything [babel-plugin-styled-components](https://github.com/styled-components/babel-plugin-styled-components) does but as a standalone solution?

Has this already been considered? Or is this something that you'd consider interesting for the future?

---

Background: 

- Remix does not expose it's compiler therefore needing [Support SSR without the babel plugin](https://github.com/styled-components/styled-components/issues/3660) - If the standalone solution would be very lightweight. It might be sensible to build it into remix. 
- Multiple issues with rollup (https://github.com/styled-components/babel-plugin-styled-components/issues/280, https://github.com/styled-components/babel-plugin-styled-components/issues/284) seem to originate in compatibility issues between rollups babel plugin and the styled components babel plugin

**Comment by quantizor** [comment]
The ecosystem has addressed this since this was posted. There are now several non-Babel alternatives:

| Tool | Package | Status |
|------|---------|--------|
| **Next.js** (built-in) | `compiler.styledComponents` in next.config.js | Active, production-ready. Supports displayName, SSR, fileName, minification, cssProp, pure annotations. Works with both Webpack and Turbopack. |
| **SWC standalone** | [`@swc/plugin-styled-components`](https://www.npmjs.com/package/@swc/plugin-styled-components) | Active (v12.7.0, Feb 2026). Usable with any SWC-based toolchain outside Next.js. |
| **Rsbuild/Rspack** | [`@rsbuild/plugin-styled-components`](https://www.npmjs.com/package/@rsbuild/plugin-styled-components) | Active (v1.6.1). Wraps the SWC plugin. |
| **Vite** | Use `@vitejs/plugin-react-swc` + `@swc/plugin-styled-components` | No dedicated plugin, but SWC integration works. |

styled-components v6 also works **without any build plugin** — the main trade-off is less informative component names in React DevTools. SSR hydration stability is generally fine with modern bundlers since module evaluation order is consistent.

The original `babel-plugin-styled-components` (v2.1.4, June 2023) still works for Babel-based setups but receives no updates.
